pact4s/src/test/scala/org/broadinstitute/dsde/rawls/provider/RawlsProviderSpec.scala

```diff
@@ -0,0 +1,243 @@
+import org.broadinstitute.dsde.rawls.spendreporting.SpendReportingService
+import org.broadinstitute.dsde.rawls.status.StatusService
+import org.broadinstitute.dsde.rawls.user.UserService
+import org.broadinstitute.dsde.rawls.webservice.RawlsApiServiceImpl
+import org.broadinstitute.dsde.rawls.workspace.{MultiCloudWorkspaceService, WorkspaceService}
+import org.broadinstitute.dsde.workbench.oauth2.OpenIDConnectConfiguration
+import org.mockito.ArgumentMatchers.{any, anyInt, anyString}
+import org.mockito.Mockito.{reset, when}
+import org.mockito.stubbing.OngoingStubbing
+import org.scalatest.BeforeAndAfterAll
+import org.scalatest.flatspec.AnyFlatSpec
+import pact4s.provider.Authentication.BasicAuth
+import pact4s.provider.{ConsumerVersionSelectors, PactSource, ProviderInfoBuilder, ProviderTags}
+import pact4s.scalatest.PactVerifier
+import pact4s.provider._
+import org.scalatest.time.SpanSugar.convertIntToGrainOfTime
+import org.scalatestplus.mockito.MockitoSugar.mock
+import org.typelevel.log4cats.StructuredLogger
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+import pact4s.provider.StateManagement.StateManagementFunction
+
+import java.lang.Thread.sleep
+import java.util.UUID
+import java.util.concurrent.TimeUnit
+import scala.concurrent.{ExecutionContext, ExecutionContextExecutor, Future}
+import scala.concurrent.duration.FiniteDuration
+
+object States {
+  val oneSnapshot = "one snapshot in the given workspace"
+}
+
+class RawlsProviderSpec extends AnyFlatSpec with BeforeAndAfterAll with PactVerifier {
+
+  implicit val loggerIO: StructuredLogger[IO] = Slf4jLogger.getLogger[IO]
+  implicit val system: ActorSystem = ActorSystem("rawlstests")
+  implicit val materializer: Materializer.type = Materializer
+  implicit val ec: ExecutionContextExecutor = ExecutionContext.global
+
+  val mockMultiCloudWorkspaceServiceConstructor: RawlsRequestContext => MultiCloudWorkspaceService =
+    _ => mock[MultiCloudWorkspaceService]
+  val mockWorkspaceServiceConstructor: RawlsRequestContext => WorkspaceService =
+    _ => mock[WorkspaceService]
+  val mockEntityServiceConstructor: RawlsRequestContext => EntityService = _ => mock[EntityService]
+  val mockUserServiceConstructor: RawlsRequestContext => UserService = _ => mock[UserService]
+  val mockGenomicsServiceConstructor: RawlsRequestContext => GenomicsService =
+    _ => mock[GenomicsService]
+  val mockSnapshotServiceConstructor: RawlsRequestContext => SnapshotService =
+    _ => mock[SnapshotService]
+  val mockSpendReportingConstructor: RawlsRequestContext => SpendReportingService =
+    _ => mock[SpendReportingService]
+  val mockBillingProjectOrchestratorConstructor: RawlsRequestContext => BillingProjectOrchestrator =
+    _ => mock[BillingProjectOrchestrator]
+  val mockBucketMigrationServiceConstructor: RawlsRequestContext => BucketMigrationService =
+    _ => mock[BucketMigrationService]
+  val mockStatusServiceConstructor: () => StatusService = mock[() => StatusService]
```

Comment:
It is important to use the singleton pattern here for rendering`StatusService` so that the constructor returns the same service object every time being called. Below is a suggestion on how to do this. You also need to examine all other service constructors to follow the same pattern.
```suggestion
lazy val mockStatusServiceConstructor: () => StatusService = {
    // Create and return a singleton mockStatusService
    lazy val mockStatusService: StatusService = mock[StatusService]
    () => mockStatusService
  }
```

Reply:
Does it make a difference whether the test is actually using or relying on these services?  Most of these are being constructed just to have something to initialize the RawlsApi with but not actually used in the test in any way.  Can I get away with only doing this for service constructors under test?

Reply:
You need to use singleton object. Think this way, when you spin up the service on the mock server using the constructor it has all the global states initiated with this copy of object.

When you use mockito to mock the behavior of endpoint, you want to mock the method referenced by the same object. When use call the constructor the second time without singleton, it will generate a new mock object and so you end up not mocking the intended copy deployed to the mock server.

Reply:
So to answer the question, you can get away with singleton for those services not under test. In my opinion though, if someone else in the future picks up some new service to be tested, they won't be puzzled.

Reply:
Another thought is how do you guarantee that a service is a standalone service and other services don't depend on, and that happen to be the service with a non-singleton constructor.